### ISSUE TITLE
api-pagination 4.5.0 support for Rails 4.2.x broken?

### ISSUE THREAD

**Issue by ShaneBoyer** [davidcelis/api-pagination#71]
The update for Rails 5 support has in some way broken support for Rails 4.2. Reverting to api-pagination 4.4.0 restores working support for Rails 4.2.  On rails server startup the following trace occurs in Rails 4.2.x with api-pagination 4.5.0:

```
Bundler::GemRequireError: There was an error while trying to load the gem 'api-pagination'.
Gem Load Error is: uninitialized constant ApiPagination::Hooks
Backtrace for gem load error is:
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/api-pagination-4.5.0/lib/api-pagination/hooks.rb:7:in `block in <top (required)>'
```

**Comment by davidcelis** [comment]
Do you have a longer stack trace, by any chance?

**Comment by ShaneBoyer** [comment]
Sorry, I expect this section should help:

```
Bundler Error Backtrace:
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:89:in `rescue in block (2 levels) in require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:85:in `block (2 levels) in require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:81:in `each'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:81:in `block in require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:70:in `each'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:70:in `require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler.rb:102:in `require'
/var/lib/jenkins/workspace/zzz/config/application.rb:13:in `<top (required)>'
/var/lib/jenkins/workspace/zzz/Rakefile:4:in `require'
/var/lib/jenkins/workspace/zzz/Rakefile:4:in `<top (required)>'
NameError: uninitialized constant ApiPagination::Hooks
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/api-pagination-4.5.0/lib/api-pagination/hooks.rb:7:in `block in <top (required)>'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/activesupport-4.2.7.1/lib/active_support/lazy_load_hooks.rb:38:in `instance_eval'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/activesupport-4.2.7.1/lib/active_support/lazy_load_hooks.rb:38:in `execute_hook'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/activesupport-4.2.7.1/lib/active_support/lazy_load_hooks.rb:28:in `block in on_load'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/activesupport-4.2.7.1/lib/active_support/lazy_load_hooks.rb:27:in `each'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/activesupport-4.2.7.1/lib/active_support/lazy_load_hooks.rb:27:in `on_load'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/api-pagination-4.5.0/lib/api-pagination/hooks.rb:6:in `<top (required)>'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/api-pagination-4.5.0/lib/api-pagination.rb:87:in `<top (required)>'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:86:in `require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:86:in `block (2 levels) in require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:81:in `each'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:81:in `block in require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:70:in `each'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler/runtime.rb:70:in `require'
/var/lib/jenkins/.rvm/gems/ruby-2.1.7/gems/bundler-1.12.5/lib/bundler.rb:102:in `require'
/var/lib/jenkins/workspace/zzz/config/application.rb:13:in `<top (required)>'
```

**Comment by davidcelis** [comment]
Do you have a require statement for `api-pagination` anywhere, or are you relying on Rails/bundler to load it for you?

Does adding `require 'api-pagination'` or `require 'api-pagination/hooks'` in an initializer or `config/application.rb` help?

I don't suppose you're able to reproduce this in a small, fresh Rails app that I could look at?

**Comment by ShaneBoyer** [comment]
Bundler.require pulls in all gems on start.  To reproduce, setup a new rails 4.2.x API project:
gem install rails-api
rails-api new api-test
cd api-test

Add the following gems to Gemfile:
gem 'kaminari'
gem 'api-pagination'

bundle install, then rails s fails immediately with the NameError.  The rails-api gem is causing the ActiveSupport.on_load hook to be called before the initial parsing of hooks.rb is completed; as a result ApiPagination::Hooks module is not yet defined since it's at the bottom of hooks.rb.  In a non-api Rails 4.2 project the ActiveSupport.on_load hook is not called until the first request is processed allowing the initialization pass to pickup the ApiPagination::Hooks definition.

The fix appears to be to move the ApiPagination::Hooks module above the on_load use of the module (perhaps at the top of the if defined?(Rails) block).  Trying this locally solved the issue for me and worked on rails 4.2.x and rails 5.x api and non-api projects.  I'm not comfortable submitting a pull request without more time getting familiar with the code or how to update the tests to cover this case.

A rails 5 api project does not (currently) have this issue due to the timing of when the ActiveSupport.on_load hook is activated, so this issue appears specific to rails 4.2 and possibly earlier using rails-api.  I didn't check rails < 4.2.

Thank you for this gem and your responsiveness!

**Comment by Dev-Crea** [comment]
I fixed this problem by specifying the version in the Gemfile.
`gem 'api-pagination', '~> 4.4', '<= 4.4.0'`

**Comment by davidcelis** [comment]
> The fix appears to be to move the ApiPagination::Hooks module above the on_load use of the module (perhaps at the top of the if defined?(Rails) block).

Interesting; I thought that _might_ solve the problem, but I had convinced myself otherwise because that `ActiveSupport.on_load` hook is supposed to be lazy. @ShaneBoyer I can patch that up later today or, since you found the bug and the fix, would you like to submit a pull request and get a contributor badge? :wink:

**Comment by ShaneBoyer** [comment]
Sure, I'll get you a pull request later today.

**Comment by davidcelis** [comment]
Closed by #72
